src/main/scala/de/unruh/isabelle/pure/ToplevelState.scala

```diff
@@ -24,6 +24,39 @@ final class ToplevelState private (val mlValue: MLValue[ToplevelState]) extends
   /** Returns the proof context corresponding to this toplevel state (ML function `Toplevel.context_of`). */
   def context(implicit isabelle: Isabelle): Context =
     Ops.getContext(this).retrieveNow
+
+  /** Returns the proof level of this state (number of opened blocks, ML function `Toplevel.level`).
+   * Zero when outside of proofs.
+   * For Skipped_Proof, this is instead the depth (plus one).
+   */
+  def proofLevel(implicit isabelle: Isabelle): Int = Ops.getProofLevel(this).retrieveNow
+
+
+  /** Returns whether the state is in top-level mode, meaning outside a theory (before "theory .. begin" or after "end"). */
+  def isTopLevelMode(implicit isabelle: Isabelle): Boolean = Ops.isTopLevelMode(this).retrieveNow
+  /** Returns whether the state is in theory mode, global or local. */
+  def isTheoryMode(implicit isabelle: Isabelle): Boolean = Ops.isTheoryMode(this).retrieveNow
+  /** Returns whether the state is in local theory mode. */
+  def isLocalTheoryMode(implicit isabelle: Isabelle): Boolean = Ops.isLocalTheoryMode(this).retrieveNow
+  /** Returns whether the state is in proof (non-skipped) mode. */
+  def isProofMode(implicit isabelle: Isabelle): Boolean = Ops.isProofMode(this).retrieveNow
+  /** Returns whether the state is in skipped-proof mode. */
+  def isSkippedMode(implicit isabelle: Isabelle): Boolean = Ops.isSkippedMode(this).retrieveNow
+  /** Returns whether the state is after a theory "end". */
+  def isEndTheory(implicit isabelle: Isabelle): Boolean = Ops.isEndTheory(this).retrieveNow
+  /** Returns the mode of the state, one of: top-level/theory/local-theory/proof/skipped-proof. */
+  def mode(implicit isabelle: Isabelle): String = {
```

Comment:
I'm not a friend of using strings for enumeration (less type safety). Can you do one of:
* Change this to return an enumeration type ToplevelState.Mode
* If the purpose of this function is actually to return a human-readable string, rename it to something like `modeString` (and mention in the comment that it's purpose is for human readable messages)
* Remove it.

Reply:
Ok, made an enum.